src/target/aby/trans.rs

```diff
@@ -0,0 +1,477 @@
+                            EmbeddedTerm::Bool(format!("circ->PutGTGate({}, {})", a, b)),
+                        );
+                    }
+                    BvBinPred::Ule => {
+                        let eq = self.embed_eq(t.clone(), &t.cs[0], &t.cs[1]);
+                        self.cache.insert(
+                            t.clone(),
+                            EmbeddedTerm::Bool(format!(
+                                "((BooleanCircuit *)circ)->PutORGate(circ->PutGTGate({}, {}), {})",
+                                b, a, eq
+                            )),
+                        );
+                    }
+                    BvBinPred::Ult => {
+                        self.cache.insert(
+                            t.clone(),
+                            EmbeddedTerm::Bool(format!("circ->PutGTGate({}, {})", b, a)),
+                        );
+                    }
+                    _ => panic!("Non-field in bool BvBinPred: {}", op),
+                }
+            }
+            _ => panic!("Non-field in embed_bool: {}", t),
+        }
+
+        self.get_bool(&t)
+    }
+
+    /// Given term `t`, type-check `t` is of type Bv and return the variable name for
+    /// `t`
+    fn get_bv(&self, t: &Term) -> String {
+        match self
+            .cache
+            .get(t)
+            .unwrap_or_else(|| panic!("Missing wire for {:?}", t))
+        {
+            EmbeddedTerm::Bv(b) => b.clone(),
+            _ => panic!("Non-bv for {:?}", t),
+        }
+    }
+
+    fn embed_bv(&mut self, t: Term) -> String {
+        match &t.op {
+            Op::Var(name, Sort::BitVector(_)) => {
+                if !self.inputs.contains_key(name) {
+                    self.inputs
+                        .insert(name.to_string(), *self.md.inputs.get(name).unwrap());
+                    self.inputs_order.push(name.to_string());
+                }
+                if !self.cache.contains_key(&t) {
+                    self.cache
+                        .insert(t.clone(), EmbeddedTerm::Bv(format!("s_{}", name)));
+                }
+            }
+            Op::Const(Value::BitVector(b)) => {
+                self.cache.insert(
+                    t.clone(),
+                    EmbeddedTerm::Bool(format!(
+                        "circ->PutCONSGate((uint64_t){}, (uint32_t){})",
+                        b,
+                        b.width()
+                    )),
+                );
+            }
+            Op::Ite => {
+                let sel = self.get_bool(&t.cs[0]).clone();
+                let a = self.get_bv(&t.cs[1]).clone();
+                let b = self.get_bv(&t.cs[2]).clone();
+                self.cache.insert(
+                    t.clone(),
+                    EmbeddedTerm::Bool(format!("circ->PutMUXGate({}, {}, {})", a, b, sel)),
+                );
+            }
+            Op::BvNaryOp(o) => {
+                let a = self.get_bv(&t.cs[0]);
+                let b = self.get_bv(&t.cs[1]);
```

Comment:
This isn't quite right. While your tests seem to only be generating _binary_ additions, multiplications, etc., these are n-ary operators, and the compiler might give you n-ary instances of them. Perhaps you need to fold over the list here?

Reply:
You should be able to create a problematic test case by just adding 3 numbers (unless I'm missing something)

Reply:
Alternatively, you might write and IR-to-IR transformation pass that just binarizes the operator. That would be more re-usable, but perhaps a bit more code.

Reply:
I added an nary_add test case:

def main(private<1> u32 a, private<2> u32 b,  private<2> u32 c) -> u32:
  return a + b + c

and I got back this set of terms:

(bvadd (bvadd main_f0_lex0_a_v0 main_f0_lex0_b_v0) main_f0_lex0_c_v0)

I checked the MPC optimizations and it looks like we removed the flatten optimization since all of the gates only take in binary inputs anyways. 

Without the flatten optimization, I think there shouldn't be any nary-terms (unless I'm also missing something here).

Reply:
Ah, I see---I was missing something. We're getting lucky here, in a sense, because the ZoKrates front-end is not producing an n-ary term for `a + b + c`.

That doesn't mean the front-end *never* produces n-ary terms though. I bet it does in some array cases.

Reply:
I see, is there any difference between these two options other than reusability?
- writing an IR-to-IR transformation that binarizes the operator 
- using a for loop to iterate and append a gate for each of the inputs

Reply:
I don't see any.

Reply:
I'll add the binarize translation layer into the IR in the next PR!